### ISSUE TITLE
About variables 'hiddens' and 'outputs'

### ISSUE THREAD

**Issue by q-rong** [hobincar/RecNet#21]
> 
> 
> (1) I saved decoder hidden states into the variable `hiddens` where the shape of a decoder hidden state is `(D=1, B=200, H=512)`. Therefore, the shape of the variable `hiddens` should be `(self.max_caption_len + 2, D, B, H)`. The shape of the variable `outputs` should be `(self.max_caption_len + 2, D, B, V=512)`. BTW, the code was hard to read. I'll modify this to be more readable.
> 
> (2) When we use LSTM as a decoder, the variable `hidden` will be a tuple of two tensors `(HIDDEN_STATE, CONTEXT_STATE)`. When we use GRU as a decoder, the variable `hidden` will be a tensor `HIDDEN_STATE`. However, we only need the `HIDDEN_STATE` for our reconstructor's input. That's why I divide the cases into LSTM / GRU.

Hi,

(1) I notice you say  `The shape of the variable 'outputs' should be '(self.max_caption_len + 2, D, B, V=512)'`, but in your code it's `outputs = Variable(torch.zeros(self.max_caption_len + 2, batch_size, vocab_size=13380)).cuda()`. 

I found that one parameter was missing. Is this an error?

(2) I don't understand why the variable `hiddens` has two tensors`(HIDDEN_STATE, CONTEXT_STATE)`.
What is `CONTEXT_STATE` and what does it do? What does `HIDDEN_STATE` have to do with `CONTEXT_STATE`?

(3) Why do you want to `bottleneck_size` set to 256, rather than the other values, such as `512` or `1536` and so on.  Reduce `self.feat_size=1536` and `self.hidden_size=512` to 256. Won't there be data loss?
```
class TemporalAttention(nn.Module):
    def __init__(self, hidden_size, feat_size, bottleneck_size):
        super(TemporalAttention, self).__init__()
        self.hidden_size = hidden_size
        self.feat_size = feat_size
        self.bottleneck_size = bottleneck_size
        self.W = nn.Linear(self.hidden_size, self.bottleneck_size, bias=False)
        self.U = nn.Linear(self.feat_size, self.bottleneck_size, bias=False)
        self.b = nn.Parameter(torch.ones(self.bottleneck_size), requires_grad=True)
        self.w = nn.Linear(self.bottleneck_size, 1, bias=False)
```

**Comment by hobincar** [comment]
Hi,

(1) Sorry. That was my mistake. the shape of `outputs` should be
```
(
   self.max_caption_len+2, # The maximun number of words including <SOS> and <EOS> tokens
   B, # The batch size
   V # the number of vocabulary. For the case of MSVD, it is 13380
),
```
since this variable denotes the probability of words in each sentence in a batch.

(2) There are two most popular types of RNNs; LSTM and GRU. LSTM maintains context using two kinds of state: a hidden state and a context state. GRU maintains context using only a hidden state. As this model uses LSTM for its decoder, the state are represented by a tuple of the two state vectors.

(3) We often reduce the size of its feature vector for represent the input. We lose something, but we can focus on some salient features and do some great job with the them. I guess the `bottleneck_size` of `256` is quite big enough, but you can increase it to whatever size you want.
